src/solver/generic_solver/treemanipulations.jl

```diff
@@ -0,0 +1,234 @@
+#TODO: tree manipulations should be callable by passing a `hole`, instead of a `path`
+# Related issue: https://github.com/orgs/Herb-AI/projects/6/views/1?pane=issue&itemId=54473456
+
+#TODO: unit test all tree manipulations
+
+"""
+    remove!(solver::GenericSolver, path::Vector{Int}, rule_index::Int)
+
+Remove `rule_index` from the domain of the hole located at the `path`.
+It is assumed the path points to a hole, otherwise an exception will be thrown.
+"""
+function remove!(solver::GenericSolver, path::Vector{Int}, rule_index::Int)
+    hole = get_hole_at_location(solver, path)
+    if !hole.domain[rule_index]
+        # The rule is not present in the domain, ignore the tree manipulation
+        return
+    end
+    hole.domain[rule_index] = false
+    simplify_hole!(solver, path)
+    notify_tree_manipulation(solver, path)
+    fix_point!(solver)
+end
+
+"""
+    remove!(solver::GenericSolver, path::Vector{Int}, rules::Vector{Int})
+
+Remove all `rules` from the domain of the hole located at the `path`.
+It is assumed the path points to a hole, otherwise an exception will be thrown.
+"""
+function remove!(solver::GenericSolver, path::Vector{Int}, rules::Vector{Int})
+    hole = get_hole_at_location(solver, path)
+    domain_updated = false
+    for rule_index ∈ rules
+        if hole.domain[rule_index]
+            domain_updated = true
+            hole.domain[rule_index] = false
+        end
+    end
+    if domain_updated
+        simplify_hole!(solver, path)
+        notify_tree_manipulation(solver, path)
+        fix_point!(solver)
+    end
+end
+
+"""
+    remove_all_but!(solver::GenericSolver, path::Vector{Int}, new_domain::BitVector)
+
+Reduce the domain of the hole located at the `path`, to the `new_domain`.
+It is assumed the path points to a hole, otherwise an exception will be thrown.
+It is assumed new_domain ⊆ domain. For example: [1, 0, 1, 0] ⊆ [1, 0, 1, 1]
+"""
+function remove_all_but!(solver::GenericSolver, path::Vector{Int}, new_domain::BitVector)
+    hole = get_hole_at_location(solver, path)
+    if hole.domain == new_domain @warn "'remove_all_but' was called with trivial arguments" return end
+    @assert is_subdomain(new_domain, hole.domain) "($new_domain) ⊈ ($(hole.domain)) The remaining rules are required to be a subdomain of the hole to remove from"
+    hole.domain = new_domain
+    simplify_hole!(solver, path)
+    notify_tree_manipulation(solver, path)
+    fix_point!(solver)
+end
+
+"""
+    remove_above!(solver::GenericSolver, path::Vector{Int}, rule_index::Int)
```

Comment:
Why do you need the notion of above and below here? Why would the grammar order matter?

Reply:
This is needed in the `Ordered` constraint, to break symmetries.
To make sure only one of `a + b` and  `b + a` is allowed, we can impose the ordering: `a<=b`